### DISCUSSION TITLE
Handling session and authentication

### DISCUSSION THREAD

**Discussion by kamilc** [reflex-dev/reflex#280] [General]
Hey all,

This project looks super dope! I know it's still early days and that auth-related components are on their way, but just to understand the architecture a bit more - how would one go about creating one herself/himself? Also, how do we handle user sessions?

Thanks!

**Comment by FHU-yezi** [comment]
Maybe we need some ways to access Cookie, LocalStorge etc, and we can use JWT or something else.

**Comment by picklelo** [comment]
Check out the [Twitter example](https://github.com/pynecone-io/pynecone-examples/tree/main/twitter) we show how you can create your own auth logic.

I'm working on a blog post for the architecture - but in summary we have a NextJS frontend and a Python backend where all the client state is held and where the state update logic is done. A FastAPI websocket sends events to the backend to process, and state updates to the frontend to render on the UI.

**Comment by cahaj** [comment]
So, I was experimenting with an auth0 authentication system and trying to implement it in my Pynecone project and came across some things I didn't understand.

First things first, I made a really simple Flask application to demonstrate what I want to achieve.

```py
from flask import Flask, redirect, session, request

DOMAIN = 'AUTH0_DOMAIN'
ID = 'CLIENT_ID'

app = Flask('')
app.secret_key = 'RANDOM 32 HEX SECRET'


@app.route('/')
def home():
    return session.get('token', "Not logged in")

@app.route('/login')
def login():
    return redirect(f"https://{DOMAIN}/authorize?response_type=code&client_id={ID}&redirect_uri=http://127.0.0.1:3000/callback")

@app.route('/callback')
def callback():
    token = request.args.get('code')
    session['token'] = token
    return redirect("/")

@app.route('/logout')
def logout():
    del session['token']
    return redirect(f"https://{DOMAIN}/v2/logout?client_id={ID}&redirectTo=http://127.0.0.1:3000/")

def run():
    app.run(host='0.0.0.0',port=3000)

if __name__ == '__main__':
    run()
```

The things I'm struggling to recreate in Pynecone are alternative for Flask's session and returning a page redirect.

I don't understand how to temporarily store the user's token for the duration of the login session in Pynecone, am I somehow supposed to do this through the State? If I am, how am I able to call the State to build a simple if statement returning the main page when the token is found and the login page when not?

About my second problem, whenever I try to return a pc.redirect() in a API route, I get `Error: Invariant: attempted to hard navigate to the same URL /login http://127.0.0.1:3000/login`. 

Im really in love with the project, just can't find much answers to my questions by myself as it is still really new, so sorry if it's something really simple that I'm missing!

Thanks!

**Comment by steamraven** [comment]
I would really like to see some authentication and authorization features as first level features.   These features are needed in any app that goes beyond a toy or demo.  And it is very hard to get this right, as the [Twitter Example](https://github.com/pynecone-io/pynecone-examples/tree/main/twitter)  shows.   It has some serious problems and shows how easy it is to miss things.

It would be nice to have a way annotate state events (including the implicit set_* events) with the authorization level.

Authentication may be able to be handled with middleware

I really think having these features would go a long way to making this project seem complete

**Comment by larsblumberg** [comment]
Reflex provides support for rolling out a custom `rx.Middleware()` which hooks into the event processing. This allows you to validate the user session before processing further your app events. So that you can log out the user in case the user session has expired.

See related Discord discussion with @masenf: https://discord.com/channels/1029853095527727165/1469722900423770336

Here's a quotation of the most important pointers he shared:

> unfortunately the reflex middleware is not really documented. but if you subclass rx.Middleware (https://github.com/reflex-dev/reflex/blob/main/reflex/middleware/middleware.py) and then call app.add_middleware (https://github.com/reflex-dev/reflex/blob/5d1d6db5bce511397b720014c71e053c4416a065/reflex/app_mixins/middleware.py#L25-L35).
> 
> the reflex middleware is not request based, but event based; it will get called for each event that the frontend queues up.
> 
> as for the hydrate/local storage issue. we actually use a middleware to perform the hydrate using a middleware (https://github.com/reflex-dev/reflex/blob/main/reflex/middleware/hydrate_middleware.py), so if you are hooking outside the middleware system, then it's possible that your code is running first. if you append a new middleware, then it should run after the hydrate, but not necessarily after the local storage or on_load events fire.

**Comment by masenf** [comment]
Here is a complete example of how you can implement event handler protection using `rx.Middleware` for a subset of classes

```python
from collections.abc import Callable
from typing import Any, TypeVar, overload

import reflex as rx
from reflex_enterprise.auth.cookie import HTTPCookie


S = TypeVar("S", bound=rx.State)
HARDCODED_PASSWORD = "secret"
DEFAULT_ALLOWED_USERS = {"masen", "thad", "nikhil", "alek"}
PROTECTED_STATES: dict[type[rx.State], set[str]] = {}


@overload
def protect_state(
    s: type[S], allowed_users: set[str] = DEFAULT_ALLOWED_USERS
) -> type[S]: ...
@overload
def protect_state(
    s: None = None, allowed_users: set[str] = DEFAULT_ALLOWED_USERS
) -> Callable[[type[S]], type[S]]: ...


def protect_state(
    s: type[rx.State] | None = None, allowed_users: set[str] = DEFAULT_ALLOWED_USERS
):
    if s is not None:
        PROTECTED_STATES[s] = allowed_users
        return s

    def _wrapper(cls: type[rx.State]):
        PROTECTED_STATES[cls] = allowed_users
        return cls

    return _wrapper


class AuthState(rx.State):
    _token: str = HTTPCookie()

    @rx.var
    def username(self) -> str:
        return self._token

    @rx.event
    def login(self, form_data: dict[str, Any]):
        if (username := form_data.get("username")) and (
            form_data.get("password") == HARDCODED_PASSWORD
        ):
            self._token = username.lower()
        else:
            return rx.toast(f"The password is always '{HARDCODED_PASSWORD}'. Try again.")

    @rx.event
    def logout(self):
        self._token = ""


@protect_state
class State(rx.State):
    @rx.event
    def do_thing_1(self):
        return rx.toast("Doing thing 1")

    @rx.event
    def do_thing_2(self):
        return rx.toast("Doing thing 2")


@protect_state(allowed_users={"masen"})
class MasenOnlyState(rx.State):
    @rx.event
    def do_thing_3(self):
        return rx.toast("Doing thing 3")


class AuthMiddleware(rx.Middleware):
    async def preprocess(self, app, state, event):
        for protected_state, allowed_users in PROTECTED_STATES.items():
            if event.name.startswith(protected_state.get_full_name()):
                auth_state = await state.get_state(AuthState)
                if auth_state._token not in allowed_users:
                    return rx.state.StateUpdate(
                        events=rx.event.fix_events(
                            [
                                rx.toast(
                                    "You are not authorized to perform this action."
                                )
                            ],
                            token=event.token,
                        ),
                        final=True,
                    )


def logged_in_view() -> rx.Component:
    return rx.vstack(
        rx.heading(f"Welcome, {AuthState.username}!"),
        rx.button("Logout", on_click=AuthState.logout),
    )


def logged_out_view() -> rx.Component:
    return rx.vstack(
        rx.heading("Welcome to Reflex!", size="9"),
        rx.form(
            rx.vstack(
                rx.input(name="username", placeholder="Username"),
                rx.input(name="password", placeholder="Password", type="password"),
                rx.button("Login", type="submit"),
            ),
            on_submit=AuthState.login,
        ),
    )


def index() -> rx.Component:
    return rx.container(
        rx.color_mode.button(position="top-right"),
        rx.vstack(
            rx.cond(
                AuthState.username,
                logged_in_view(),
                logged_out_view(),
            ),
            rx.button("Do Thing 1", on_click=State.do_thing_1),
            rx.button("Do Thing 2", on_click=State.do_thing_2),
            rx.button("Do Thing 3", on_click=MasenOnlyState.do_thing_3),
        ),
    )


app = rx.App()
app.add_middleware(AuthMiddleware())
app.add_page(index)
```
